### PR TITLE
Support for dynamic horizontal scaling of modules

### LINKED ISSUE #422: Support for dynamic horizontal scaling of modules
Currently, for a module to process multiple inputs to produce multiple outputs, it is required that the `Process()` method of a module implement it's own ability to do so (say, via threading to process in parallel.) However, this becomes a problem if you have a previous module being able to run multiple times, and want to send it's output via container to a latter module, that is not able to support multiple inputs. 

For example, consider a recipe that takes a snapshot of a GCP disk and sends the result to `TurbiniaGCPProcessor`. It makes a lot of sense for the GCP disk snapshotter to multiple disks as input and snapshot each of them, but `TurbiniaGCPProcessor` can only process one at a time. It would be great if `TurbiniaGCPProcessor` could be run in parallel, once for each input disk.

A potential implementation might look like:

* Module A generates a new container holding the output for each input (in our example, one container per disk snapshot)
* Module B is a subclass of a new Module base class (say, ThreadedModule.) ThreadedModule has a static method StaticSetUp - Detailed below. 
* Module B lists Module A in `"wants"` in the recipe
* If DFTW recognises that Module B is a ThreadedModule, DFTW runs the `StaticSetUp()` method of the module
* DFTW launches a copy of Module B in parallel, providing it one container that Module A generated

Ideally, this would be designed such that existing modules require no updates to be able to take advantage of this. 

The point of `StaticSetUp` as part of `ThreadedModule` is so actions that need to be taken only once no matter how many modules are run can be placed there. For example, in a module that creates multiple disk snapshots in GCP, `StaticSetUp` could be used to create the IAM permissions required in a GCP account.

### PR DESCRIPTION
**Pull request by ramo-j**
Implement a method of dynamically threading modules to process in parallel. This is done with a ThreadAwareModule that `state.py` is able to identify and then launch parallel threads for, providing input for each via the container store.

How to implement a ThreadAwareModule copied from inline docs:

```
  ThreadAwareModule are modules designed to handle being run in parallel.

  How to implement this class:

   * This module will be threaded based on a nominated ThreadOn container type:
  so nominated by returning the class type in GetThreadOnContainerType().

  * There are a number of static methods that you can use to run once,
  regardless of the number of threads that will run.

  * SetUp will run once.

  * Process will run N times in GetThreadPoolSize() parallel threads - N is the
  number of containers of the nominated type generated by previous modules.
  Process will be passed one container of the type specified by
  GetThreadOnContainerType().
```

Closes #422

### CODE REVIEW

**Comment by ramo-j** [comment]
Note from external review:

```
ThreadAwareModule.get/storecontainer looks VERY similar to ThreadAwareModule.state.get/storecontainer
they could easily be confused..if someone uses the wrong one will it be a probelm? maybe it should be get/set threadedContainer 
```

**Review by tomchop — CHANGES_REQUESTED** [review]
This change is on the right track, but I think it needs some more testing for some edge cases, namely a module that:

* Stores different types of containers (different, or not, than the container type it threads on)
* Gets containers from the state (without popping), and adds other containers of the same type to the state (effectively increasing the total number of containers of the same type)

I am still a bit uneasy with the deepcopying of the state and the module (it makes things harder to debug). I appreciate that this change makes life simpler for module developers (the idea being they just have to inherit from a class and not migrate all their code), but right now there are too many edge cases that will break existing modules. I also think explicit is better than implicit - I'd rather have module developers write their logic with threading in mind (even if it implies moderate refactoring of their code) than benefit from magical paralellization that hopefully works on their module.

I maintain that a simpler and more explicit implementation could be achieved by adding a new `HorizontalProcess` function (that takes in a list of containers) in the module interface that modules that wish to implement horizontal scaling can move their processing logic to. This function would be threaded at the state level and take in a single container as a parameter (that you can think of as the store for that individual thread).

I'm happy to discuss this in more details. I think it would require marginally more refactoring for modules and it has the advantage of not clearly exposing the execution logic to module developers.

File: dftimewolf/lib/module.py [hunk 1]
```diff
@@ -137,3 +146,143 @@ def SetUp(self, *args, **kwargs) -> None:  # type: ignore[no-untyped-def]
...
+    self._thread_lock = threading.Lock()
+    self.store = {}  # type: Dict[str, List[interface.AttributeContainer]]
+
+  def __deepcopy__(self, memo: Dict[Any, Any]) -> object:
+    """Override of deepcopy. We cheat a little - The container to thread on is
+    deepcopy'd, but other containers are shallow copied - so all instances of
+    the module can access and modify them by reference"""
+    state = deepcopy(self.state, memo)
+    copy = ThreadAwareModule(state) # type: ignore
+    copy._thread_lock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/module.py:202]: Nah, you're correct, it's creating a lock twice unnecessarily.

File: dftimewolf/lib/module.py [hunk 2]
```diff
@@ -137,3 +146,143 @@ def SetUp(self, *args, **kwargs) -> None:  # type: ignore[no-untyped-def]
...
+    copy.__class__ = type(self)
+
+    # Deep copy the containers to thread on, shallow copy the rest.
+    for key, container_list in self.store.items():
```
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/module.py:206]: Nah these get populated into the modules `store` in `state.py`

File: dftimewolf/lib/module.py [hunk 3]
```diff
@@ -137,3 +146,143 @@ def SetUp(self, *args, **kwargs) -> None:  # type: ignore[no-untyped-def]
...
+  # this module should use these methods to access containers, rather than
+  # the same methods in the state.
+  def StoreContainer(self, container: "interface.AttributeContainer") -> None:
+    """Thread-safe method to store data in the state's store.
+
+    Args:
+      container (AttributeContainer): data to store.
+    """
+    with self._thread_lock:
+      self.store.setdefault(container.CONTAINER_TYPE, []).append(container)
```
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/module.py:267]: I left it in from the copied version, with the thought that each module might still implement it's own threading within itself.

File: dftimewolf/lib/state.py [hunk 4]
```diff
@@ -60,6 +63,27 @@ def __init__(self, config: Type[Config]) -> None:
     self.streaming_callbacks = {}  # type: Dict[Type[interface.AttributeContainer], List[Callable[[Any], Any]]]  # pylint: disable=line-too-long
     self._abort_execution = False
 
+  def __deepcopy__(self, memo: Dict[Any, Any]) -> object:
+    """Deepcopy override. This creates a new copy of the state object. This is
+    needed because we copy ThreadAwareModules for parallel processing, and they
+    have a reference to the state."""
```
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/state.py:69]: The deepcopy was necessary because the thread lock in the global state cannot be deepcopy'd - So the deepcopy override creates an individual thread lock.

File: dftimewolf/lib/state.py [hunk 5]
```diff
@@ -308,7 +336,51 @@ def _RunModuleThread(self, module_definition: Dict[str, str]) -> None:
...
+        for _, container_list in self.store.items():
+          for container in container_list:
+            module.StoreContainer(container)
+
+        module.StaticPreProcess()
+
+        # Thread Aware Modules must have only one copy of the container that
+        # they thread based on. Pop them.
+        thread_on_containers = module.GetContainers(
+            module.GetThreadOnContainerType(), True)
```
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/state.py:351]: Yep, cleaned up the double handling of those containers.

File: dftimewolf/lib/state.py [hunk 6]
```diff
@@ -308,7 +336,51 @@ def _RunModuleThread(self, module_definition: Dict[str, str]) -> None:
...
+
+        for fut in futures:
+          if fut.exception():
+            raise fut.exception() # type: ignore
+
+        # Collect any output containers
+        for m in modules:
+          for return_container in \
+              m.GetContainers(module.GetThreadOnContainerType(), True):
+            return_containers.append(return_container)
```
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/state.py:372]: Yep, good catch. Updated this to copy any new containers we didn't have originally, and updated the container handling test to confirm that works.

**Comment by ramo-j** [comment]
The design of this I sense is going to be a trade off between the simplification of container handling, vs effort to implement the module, and upgrading existing modules. 

Updated this with a new balance that I think is reasonable.

File: dftimewolf/lib/module.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/module.py]: Removed with updated container handling.

File: dftimewolf/lib/state.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/state.py]: Removed with updated container handling.

File: dftimewolf/lib/module.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ramo-j) [reply dftimewolf/lib/module.py]: Removed with updated container handling.

**Review by tomchop — APPROVED** [review]
LGTM, thanks a lot for the changes!
